supportedBackends/aws/src/main/scala/cromwell/backend/impl/aws/AwsBatchJob.scala

```diff
@@ -164,79 +222,156 @@ final case class AwsBatchJob(jobDescriptor: BackendJobDescriptor, // WDL/CWL
-      case _ => false
-    }).compile.last.map(_.get)
-
-    async.recoverWith(retry) {
-      case e: ClientException if e.statusCode() == 409 =>
-        // This could be a problem here, as we might have registerJobDefinition
-        // but not describeJobDefinitions permissions. We've changed this to a
-        // warning as a result of this potential
-        Log.warn("Job definition already exists. Performing describe and retrieving latest revision.")
-        async.
-          delay(client.describeJobDefinitions(DescribeJobDefinitionsRequest.builder().jobDefinitionName(sanitize(name)).build())).
-          map(_.jobDefinitions().asScala).
-          map(_.last.jobDefinitionArn())
+  private def findOrCreateDefinition[F[_]]()
+                                    (implicit async: Async[F], timer: Timer[F]): Aws[F, String] = ReaderT { awsBatchAttributes =>
+
+    // this is a call back that is executed below by the async.recoverWithRetry(retry)
+    val submit = async.delay({
+
+      val commandStr = awsBatchAttributes.fileSystem match {
+        case AWSBatchStorageSystems.s3 => reconfiguredScript
+        case _ => script
+      }
+      val jobDefinitionContext = AwsBatchJobDefinitionContext(
+        runtimeAttributes = runtimeAttributes,
+        commandText = commandStr,
+        dockerRcPath = dockerRc,
+        dockerStdoutPath = dockerStdout,
+        dockerStderrPath = dockerStderr,
+        jobDescriptor = jobDescriptor,
+        jobPaths = jobPaths,
+        inputs = inputs,
+        outputs = outputs)
+
+      val jobDefinitionBuilder = StandardAwsBatchJobDefinitionBuilder
+      val jobDefinition = jobDefinitionBuilder.build(jobDefinitionContext)
+
+
+      //check if there is already a suitable definition based on the calculated job definition name
+      val jobDefinitionName = jobDefinition.name
+
+      Log.info(s"Checking for existence of job definition called: $jobDefinitionName")
+
+      val describeJobDefinitionRequest = DescribeJobDefinitionsRequest.builder()
+        .jobDefinitionName( jobDefinitionName )
+        .status("ACTIVE")
+        .build()
+
+      val describeJobDefinitionResponse = batchClient.describeJobDefinitions(describeJobDefinitionRequest)
+
+      if ( !describeJobDefinitionResponse.jobDefinitions.isEmpty ) {
+
+        Log.info(s"Found job definition $jobDefinitionName, getting arn for latest version")
+
+        //sort the definitions so that the latest revision is at the head
+        val definitions = describeJobDefinitionResponse.jobDefinitions().asScala.toList.sortWith(_.revision > _.revision)
+
+        Log.info(s"Latest job definition revision is: ${definitions.head.revision()} with arn: ${definitions.head.jobDefinitionArn()}")
+
+        //return the arn of the job
+        definitions.head.jobDefinitionArn()
+      } else {
+
+        //no definition found. create one
+        Log.info(s"No job definition found")
+
+        Log.info(s"Creating job definition: $jobDefinitionName")
```

Comment:
Could these two log messages be combined?

Reply:
Yes! Done.